### ISSUE TITLE
BIND(F)  (analog to BIND(C))

### ISSUE THREAD

**Issue by tclune** [j3-fortran/fortran_proposals#269]
I have a project that is starting to heavily use shared object libraries (aka DSOs’).   The interfaces are all in Fortran.   To use the DSO we must pass the global name of the called procedure to dlsym().      One might hope that declaring the DSO procedure with BIND(C) would allow us to specify the global name in a portable manner, but … the interface is not C interoperable.   (In theory it might be made so, but the underlying framework is huge and no-one is going to rewrite it to use C interoperable types all the way down.)
 
We’ve mitigated the name-mangling by putting the main DSO procedure outside of the module that would otherwise be its proper home.   At the moment most/all extant compilers with default compilation flags will then make a global name that is the lower case version of the Fortran name with an underscore.  So at least this is nominally portable, but not guaranteed to be.
 
It seems a natural request then to allow a procedure to be declared BIND(F, name=”…”)   The procedure could then even live inside of a module (amusingly allowing the module to have no PUBLIC entities).

**Comment by klausler** [comment]
`BIND(NAME="foo")` or `BIND(FORTRAN, NAME="foo")` would be better spellings than `F`.

**Comment by jeffhammond** [comment]
Is the generated binding specific to the Fortran compiler used, or is it, like `BIND(C..`, portable within that platform ABI?

**Comment by tclune** [comment]
@jeffhammond to be portable across compilers I suspect one will be right back at the requirements that are about as strict as those for BIND(C), albeit possibly spelled a bit differently.    Seems like BIND(C) is already a solution for "full" portability, or as much as is possible module compiler switches and such. 

@klausler I agree that `Fortran` is a better spelling than `F` for the standard,  but for this discussion `BIND(F)` seems to be bette for grabbing attention.  (Nod to `F--` of old ...)

**Comment by ashe2** [comment]
If this only specifies the binding label of the procedure and nothing about calling conventions, `BIND(NAME="foo")` makes more sense.

**Comment by ivan-pi** [comment]
C++ developers have also gotten sick of dealing with Fortran's different ABI's. There is a proposal to standardize BLAS in C++, motivated in part by shortcomings in Fortran, and having to maintain code to deduce which Fortran ABI is used. For more details see: 
- [P1674R2: Evolving a Standard C++ Linear Algebra Library from the BLAS](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2022/p1674r2.html)
- [P1673R3: A free function linear algebra interface based
on the BLAS](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2021/p1673r3.pdf)

Fortran's ABI problems where also mentioned in the following episode of CppCast: https://cppcast.com/spack/. Starting at about 54:38, @tgamblin says: 
> "Fortran is worse about ABI than C++. ... It gets pretty nasty."

**Comment by perazz** [comment]
I agree. That's one of the reasons I thought Fortran should try to share ABI for at least a subset of the language with C++ rather than other high-level languages, see [https://github.com/j3-fortran/fortran_proposals/issues/261#issuecomment-1167338640](this discussion)

**Comment by ivan-pi** [comment]
> The interesting ABI problems in Fortran come from array descriptors and other nontrivial things.

It's a shame the C-Fortran array descriptor is also incompatible among different implementations: https://github.com/j3-fortran/fortran_proposals/issues/247
